File: apps/ios/LanjingQuiz/Networking/ExamHTMLParser.swift
import Foundation

struct ExamParseResult: Equatable {
    let examResultsId: String?
    let examInfoId: String
    let uuid: String?
    let testIds: [String]
    let questionStates: [QuestionState]
    let sectionMap: [String: SectionStats]
    let sectionOrder: [String]
}

/// Port of apps/web/server.js parseExamHtml. Works on NSString so all match ranges are
/// UTF-16 — identical indexing semantics to the JS string operations.
enum ExamHTMLParser {

    static func parse(_ html: String, fallbackExamInfoId: String, knownResultsId: String? = nil) -> ExamParseResult {
        let ns = html as NSString
        let fullRange = NSRange(location: 0, length: ns.length)

        func extractVar(_ name: String) -> String? {
            guard let regex = try? NSRegularExpression(pattern: "var\\s+\(name)\\s*=\\s*['\"]([^'\"]+)['\"]"),
                  let match = regex.firstMatch(in: html, range: fullRange)
            else { return nil }
            return ns.substring(with: match.range(at: 1))
        }

        let examResultsId = knownResultsId ?? extractVar("exam_results_id")
        let examInfoId = extractVar("exam_info_id") ?? fallbackExamInfoId

        // Section titles with their positions (server: sectionMatches). The
        // class attr tolerates a trailing space / extra classes (资料分析 comb
        // sections are emitted as `<div class="card-content-title ">`).
        let sectionBounds: [(title: String, pos: Int)] = {
            guard let regex = try? NSRegularExpression(pattern: "<div class=\"([^\"]*card-content-title[^\"]*)\">([^<]+)</div>") else { return [] }
            return regex.matches(in: html, range: fullRange).map { match in
                (title: ns.substring(with: match.range(at: 2)), pos: match.range.location)
            }
        }()

        // Comb (资料分析) groups: an insert-list div wraps several sub-questions
        // and carries the combId in its own questionsId attribute — the payload
        // /exam/get_question_info/ needs per question to fetch the shared
        // parent_info. Depth-tracking finds where each insert-list div closes,
        // so only cards actually inside a comb inherit its combId (regular
        // cards that follow a comb section must not) — server: combBounds.
        let combBounds: [(combId: String, pos: Int, end: Int)] = {
            guard let openRegex = try? NSRegularExpression(pattern: "<div class=\"([^\"]*insert-list[^\"]*)\"[^>]*questionsId=\"([^\"]+)\""),
                  let tagRegex = try? NSRegularExpression(pattern: "<div\\b[^>]*>|</div>")
            else { return [] }
            let opens = openRegex.matches(in: html, range: fullRange).map { match in
                (combId: ns.substring(with: match.range(at: 2)).trimmingCharacters(in: .whitespacesAndNewlines),
                 pos: match.range.location,
                 end: match.range.location)
            }
            guard !opens.isEmpty else { return [] }
            var combs = opens
            let indexByPos = Dictionary(uniqueKeysWithValues: combs.enumerated().map { ($0.element.pos, $0.offset) })
            var depth = 0
            var pendingIndex: Int?
            var pendingDepth = 0
            for tag in tagRegex.matches(in: html, range: fullRange) {
                let text = ns.substring(with: tag.range)
                if text.hasPrefix("</") {
                    depth -= 1
                    if let pending = pendingIndex, depth == pendingDepth {
                        combs[pending].end = tag.range.location + tag.range.length
                        pendingIndex = nil
                    }
                } else {
                    if let index = indexByPos[tag.range.location] {
                        pendingIndex = index
                        pendingDepth = depth
                    }
                    depth += 1
                }
            }
            return combs
        }()

        // Card chunks = text between consecutive anchors (server: html.split(/<a\s+href="#[^"]*">\s*/), skipping chunk 0)
        guard let anchorRegex = try? NSRegularExpression(pattern: "<a\\s+href=\"#[^\"]*\">\\s*") else {
            return ExamParseResult(examResultsId: examResultsId, examInfoId: examInfoId, uuid: nil,
                                   testIds: [], questionStates: [], sectionMap: [:], sectionOrder: [])
        }
        let anchors = anchorRegex.matches(in: html, range: fullRange)

        var states: [QuestionState] = []
        var seen = Set<String>()

        for (i, anchor) in anchors.enumerated() {
            let chunkStart = anchor.range.location + anchor.range.length
            let chunkEnd = (i + 1 < anchors.count) ? anchors[i + 1].range.location : ns.length
            guard chunkStart < chunkEnd else { continue }
            let chunk = ns.substring(with: NSRange(location: chunkStart, length: chunkEnd - chunkStart)) as NSString
            let chunkRange = NSRange(location: 0, length: chunk.length)

            guard let qIdMatch = firstMatch("questionsId=\"([^\"]+)\"", in: chunk, range: chunkRange) else { continue }
            let qId = chunk.substring(with: qIdMatch.range(at: 1)).trimmingCharacters(in: .whitespacesAndNewlines)
            if qId.isEmpty || seen.contains(qId) { continue }
            seen.insert(qId)

            let uId = firstMatch("uuId=\"([^\"]+)\"", in: chunk, range: chunkRange)
                .map { chunk.substring(with: $0.range(at: 1)).trimmingCharacters(in: .whitespacesAndNewlines) }

            // Raw number text: comb sub-questions use "1.1"…"15.5" style labels,
            // ordinary questions a plain integer — keep the string as-is.
            let num: String = {
                guard let match = firstMatch(">\\s*(\\d+(?:\\.\\d+)?)\\s*</span>", in: chunk, range: chunkRange) else { return "" }
                return chunk.substring(with: match.range(at: 1)).trimmingCharacters(in: .whitespacesAndNewlines)
            }()

            let (state, marked): (QuestionState.State, Bool) = {
                guard let match = firstMatch("<div\\b[^>]*class=[\"']([^\"']*\\bquestion_cbox\\b[^\"']*)[\"'][^>]*>", in: chunk, range: chunkRange)
                else { return (.unanswered, false) }
                let classes = Set(chunk.substring(with: match.range(at: 1)).split(separator: " ").map(String.init))
                let state: QuestionState.State = classes.contains("right") ? .right : classes.contains("error") ? .error : .unanswered
                return (state, classes.contains("marked"))
            }()

            // Section attribution: server uses html.indexOf on the full string
            let cardPos = (html as NSString).range(of: "questionsId=\"\(qId)\"").location
            var section = ""
            for bound in sectionBounds.reversed() where cardPos > bound.pos {
                section = bound.title
                break
            }

            // The card belongs to the comb whose insert-list div actually
            // wraps it (regular cards after a comb section are outside every
            // comb range).
            var combId: String?
            for bound in combBounds where cardPos > bound.pos && cardPos < bound.end {
                combId = bound.combId
                break
            }

            states.append(QuestionState(
                questionsId: qId, uuId: uId, num: num, combId: combId, section: section,
                state: state, marked: marked
            ))
        }

        let testIds = states.map(\.questionsId)
        let uuid = states.first?.uuId ?? extractVar("uuId")

        // Per-section breakdown (server: sectionMap), "(无分类)" for empty section
        var sectionMap: [String: SectionStats] = [:]
        var sectionOrder: [String] = []
        for q in states {
            let key = q.section.isEmpty ? "(无分类)" : q.section
            if sectionMap[key] == nil { sectionOrder.append(key) }
            var stats = sectionMap[key] ?? SectionStats()
            stats.total += 1
            switch q.state {
            case .right: stats.right += 1
            case .error: stats.error += 1
            case .unanswered: stats.unanswered += 1
            }
            sectionMap[key] = stats
        }

        return ExamParseResult(
            examResultsId: examResultsId,
            examInfoId: examInfoId,
            uuid: uuid,
            testIds: testIds,
            questionStates: states,
            sectionMap: sectionMap,
            sectionOrder: sectionOrder
        )
    }

    private static func firstMatch(_ pattern: String, in string: NSString, range: NSRange) -> NSTextCheckingResult? {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        return regex.firstMatch(in: string as String, range: range)
    }
}

File: apps/android/LanjingQuiz/src/main/java/com/qzh/lanjingquiz/Domain/ExamHtmlParser.kt
package com.qzh.lanjingquiz.Domain

/**
 * 答题卡 HTML 解析 —— iOS ExamHtmlParser.swift 逐行移植(正则来自 apps/web/server.js parseExamHtml)。
 *
 * 卡片:questionsId/uuId/number/section/state/marked;state ∈ "unanswered"|"right"|"error";
 * 空 section → "(无分类)"(sectionOrder 键);组合题 insert-list 容器携带 combId。
 */
data class CardInfo(
    val questionsId: String,
    val uuId: String?,
    /** raw 题号:组合子题 "1.1" 原样保留。 */
    val number: String,
    val section: String,
    val state: String,
    val marked: Boolean,
    /** 资料分析组合题分组 id(insert-list 容器);常规题 null。 */
    val combId: String? = null,
)

data class ExamPage(
    /** 缺失时为 ""(iOS 为 nil;ApiClient.enterExam 已保证非空才进入流程)。 */
    val examResultsId: String,
    val examInfoId: String,
    val uuid: String?,
    val sections: List<String>,
    val cards: List<CardInfo>,
)

object ExamHtmlParser {

    fun parse(html: String, fallbackExamInfoId: String, knownResultsId: String? = null): ExamPage {
        val examResultsId = knownResultsId ?: extractVar("exam_results_id", html) ?: ""
        val examInfoId = extractVar("exam_info_id", html) ?: fallbackExamInfoId

        // section 标题(容忍尾随空格/多余 class:资料分析 comb section 输出 `<div class="card-content-title ">`)
        val sectionBounds = Regex("""<div class="([^"]*card-content-title[^"]*)">([^<]+)</div>""")
            .findAll(html).map { m -> m.groupValues[2] to m.range.first }.toList()

        // 组合题区间:insert-list div 携带自身 questionsId(combId);深度追踪找到闭合位置,
        // 仅真正被 insert-list 包裹的卡片继承 combId(comb section 后的常规卡片不得继承)。
        val combBounds = buildCombBounds(html)

        // 卡片块 = 相邻锚点之间的文本(server: html.split(/<a\s+href="#[^"]*">\s*/),跳过第 0 块)
        val anchors = Regex("""<a\s+href="#[^"]*">\s*""").findAll(html).map { it.range }.toList()

        val cards = mutableListOf<CardInfo>()
        val seen = mutableSetOf<String>()

        for ((i, anchor) in anchors.withIndex()) {
            // 锚点匹配含尾部空白;chunk 起点 = 匹配结束(不含)
            val chunkStart = anchor.last + 1
            val chunkEnd = if (i + 1 < anchors.size) anchors[i + 1].first else html.length
            if (chunkStart >= chunkEnd) continue
            val chunk = html.substring(chunkStart, chunkEnd)

            val qIdMatch = Regex("""questionsId="([^"]+)"""").find(chunk) ?: continue
            val qId = qIdMatch.groupValues[1].trim()
            if (qId.isEmpty() || !seen.add(qId)) continue

            val uuId = Regex("""uuId="([^"]+)"""").find(chunk)?.groupValues?.get(1)?.trim()

            // raw 题号:组合子题 "1.1"…"15.5",常规题纯整数 —— 原样保留
            val num = Regex(""">\s*(\d+(?:\.\d+)?)\s*</span>""").find(chunk)?.groupValues?.get(1)?.trim() ?: ""

            val stateMatch = Regex("""<div\b[^>]*class=["']([^"']*\bquestion_cbox\b[^"']*)["'][^>]*>""").find(chunk)
            val classes = stateMatch?.groupValues?.get(1)?.split(" ")?.map { it.trim() }?.toSet() ?: emptySet()
            val state = when {
                "right" in classes -> QuizLogic.STATE_RIGHT
                "error" in classes -> QuizLogic.STATE_ERROR
                else -> QuizLogic.STATE_UNANSWERED
            }
            val marked = "marked" in classes

            // section 归属:最后一个位于卡片之前的 section 标题(server 用全文 indexOf)
            val cardPos = html.indexOf("questionsId=\"$qId\"")
            var section = ""
            for ((title, pos) in sectionBounds.asReversed()) {
                if (cardPos > pos) { section = title; break }
            }

            var combId: String? = null
            for (comb in combBounds) {
                if (cardPos > comb.pos && cardPos < comb.end) { combId = comb.combId; break }
            }

            cards += CardInfo(qId, uuId, num, section, state, marked, combId)
        }

        // 每 section 统计与顺序("(无分类)" 占位空 section)
        val sections = mutableListOf<String>()
        for (c in cards) {
            val key = if (c.section.isEmpty()) "(无分类)" else c.section
            if (sections.none { it == key }) sections += key
        }

        val uuid = cards.firstOrNull()?.uuId ?: extractVar("uuId", html)
        return ExamPage(examResultsId, examInfoId, uuid, sections, cards)
    }

    private fun extractVar(name: String, html: String): String? =
        Regex("""var\s+$name\s*=\s*['"]([^'"]+)['"]""").find(html)?.groupValues?.get(1)

    private data class CombBound(val combId: String, val pos: Int, var end: Int)

    private fun buildCombBounds(html: String): List<CombBound> {
        val openRegex = Regex("""<div class="([^"]*insert-list[^"]*)"[^>]*questionsId="([^"]+)"""")
        val tagRegex = Regex("""<div\b[^>]*>|</div>""")
        val opens = openRegex.findAll(html).map { m ->
            CombBound(m.groupValues[2].trim(), m.range.first, m.range.first)
        }.toList()
        if (opens.isEmpty()) return emptyList()
        val combs = opens.toMutableList()
        val indexByPos = combs.mapIndexed { i, c -> c.pos to i }.toMap()
        var depth = 0
        var pendingIndex: Int? = null
        var pendingDepth = 0
        for (tag in tagRegex.findAll(html)) {
            val text = tag.value
            if (text.startsWith("</")) {
                depth -= 1
                val pending = pendingIndex
                if (pending != null && depth == pendingDepth) {
                    combs[pending].end = tag.range.last + 1   // 独占结尾,同 iOS
                    pendingIndex = null
                }
            } else {
                val index = indexByPos[tag.range.first]
                if (index != null) {
                    pendingIndex = index
                    pendingDepth = depth
                }
                depth += 1
            }
        }
        return combs
    }
}
